lib/expand/join.go

```diff
@@ -135,18 +138,19 @@ func (p *Peer) Run(listener net.Listener) error {
 	}
 	// Stopping is on best-effort basis, the client will be trying to stop the service
 	// if notified
+	p.WithField("exit-error", err).Info("Exit with error.")
 	p.stop()
-	if installpb.IsAbortedError(err) {
-		if err := p.leave(); err != nil {
-			p.WithError(err).Warn("Failed to leave cluster.")
-		}
-	}
-	if err != nil && !installpb.IsCompletedError(err) {
-		if err2 := p.fail(err.Error()); err2 != nil {
-			p.WithError(err2).Warn("Failed to mark operation as failed.")
+	if err != nil {
+		if installpb.IsAbortedError(err) {
+			if err := p.leave(); err != nil {
+				p.WithError(err).Warn("Failed to leave cluster.")
+			}
+		} else if !installpb.IsCompletedError(err) {
+			if err2 := p.fail(err.Error()); err2 != nil {
```

Comment:
Nit: I think you don't necessarily need to have to use "err2" as it won't overshadow the original err until after p.fail returns, so `err := p.fail(err.Error()); err != nil` should work fine too.